utils/cloudinfo/cloudinfo.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudinfo
+
+import (
+	"net/http"
+)
+
+type CloudProvider int
+
+const (
+	GCE CloudProvider = iota
+	AWS
+	BAREMETAL
+	UNKNOWN
+)
+
+const NOINSTANCE = "NA"
+
+type CloudInfo interface {
+	GetCloudProvider() (CloudProvider, error)
+	GetInstanceType(CloudProvider) (string, error)
+}
+
+type realCloudInfo struct{}
+
+func NewRealCloudInfo() (CloudInfo, error) {
+	return &realCloudInfo{}, nil
+}
+
+func (self *realCloudInfo) GetCloudProvider() (CloudProvider, error) {
```

Comment:
I'd suggest we split this method into two:

a private detect() method that detects which cloudprovider the node is running on. This can be called once in NewRealCloudInfo() and we can save the result as this isn't going to change across the machine's lifetime.

The public part of the interface should just return the detected type rather than calling detection logic on every invocation. We can also cache instance type as that's mostly the same detect call.